File: medium/ContainerWithMostWater.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int maxArea(vector<int> &heights) {
    size_t size = heights.size();
    int l = 0, r = size - 1;
    int mx_area = 0;
    while (l < r) {
      int area = (r - l) * min(heights[l], heights[r]);
      if (heights[l] <= heights[r]) {
        l++;
      } else {
        r--;
      }
      mx_area = max(mx_area, area);
    }
    return mx_area;
  }
};

int main() {
  Solution *s = new Solution;
  vector<int> v = {1, 7, 2, 5, 4, 7, 3, 6};
  cout << s->maxArea(v);
}

File: ts/medium/ContainerWithMostWater.js
class Solution {
	/**
	 * @param {number[]} heights
	 * @return {number}
	 */
	maxArea(heights) {
		let [left, right] = [0, heights.length - 1];
		let mx_area = 0;
		while (left < right) {
			const curWidth = right - left;
			const curHeight = Math.min(heights[left], heights[right]);
			const curArea = curWidth * curHeight;
			mx_area = Math.max(mx_area, curArea);
			heights[left] < heights[right] ? left++ : right--;
		}
		return mx_area;

	}
}
